tls_main.c

```diff
@@ -567,42 +580,74 @@ static int do_tls_setsockopt_conf(struct sock *sk, char __user *optval,
 	return rc;
 }
 
-#define TLS_LUA 99
-#define lua_error(msg) pr_warn("[lua] %s - %s\n", __func__, msg);
-
-static int do_tls_loadlua(char __user *optval, unsigned int optlen)
+static int do_tls_loadlua(struct sock *sk, char __user *optval,
+			  unsigned int optlen)
 {
+	struct tls_context *ctx = tls_get_ctx(sk);
+	struct tls_lua_info *lua_info;
+	struct tls_lua_info *opt_info;
+	char* k_init;
+	char* k_recv;
 	int rc = 0;
-	char *file;
-	lua_State *L;
 
-	file = kmalloc(optlen, GFP_KERNEL);
-	if (!file) {
-		lua_error("no memory");
-		return -ENOMEM;
+	if (!optval || (optlen < sizeof(*lua_info))) {
+		rc = -EINVAL;
+		goto out;
 	}
-	rc = copy_from_user(file, optval, optlen);
+
+	lua_info = &ctx->lua_info;
+	opt_info = (struct tls_lua_info *)optval;
+
+	/* allow reconfigure for lua script, free the former space */
+	kfree(lua_info->init);
+	kfree(lua_info->recv);
+	rc = copy_from_user(lua_info, opt_info,sizeof(*lua_info));
 	if (rc) {
 		rc = -EFAULT;
-		goto end;
+		goto out;
 	}
 
-	L = luaL_newstate();
-	if (L == NULL) {
-		lua_error("no memory");
+	/* malloc new space and swap */
+	k_init = kmalloc(lua_info->init_len, GFP_KERNEL);
+	if (!k_init) {
+		TLS_LUA_ERROR("no memory");
 		return -ENOMEM;
 	}
-	luaL_openlibs(L);
+	k_recv = kmalloc(lua_info->recv_len, GFP_KERNEL);
+	if (!k_recv) {
+		TLS_LUA_ERROR("no memory");
+		return -ENOMEM;
+	}
+	rc = copy_from_user(k_init, lua_info->init, lua_info->init_len);
+	if (rc) {
+		rc = -EFAULT;
+		goto out;
+	}
+	rc = copy_from_user(k_recv, lua_info->recv, lua_info->recv_len);
+	if (rc) {
+		rc = -EFAULT;
+		goto out;
+	}
+	lua_info->init = k_init;
+	lua_info->recv = k_recv;
 
-	if (luaL_dostring(L, file)) {
-		lua_error(lua_tostring(L, -1));
+	if (!ctx->L) {
+		ctx->L = luaL_newstate();
+		if (!ctx->L) {
+			TLS_LUA_ERROR("no memory");
+			return -ENOMEM;
+		}
+		luaL_openlibs(ctx->L);
+	}
+
+	if (luaL_dostring(ctx->L, lua_info->init)) {
```

Comment:
OK, you load and run the Lua chunk. We are missing entry points here.
The scripts will look like this:
```Lua
function process(pkt)
  -- recv() entry point
  -- do stuff with the pkt...
end

-- the global scope runs when the script is loaded
-- we should only setup stuff
setup()
print("Finished setup")
```

The `setsockopt()` runs the global scope.
The `recv()` runs the entry point.